collections/binary_search_tree.ts

```diff
@@ -255,13 +252,13 @@ export class BinarySearchTree<T> implements Iterable<T> {
           replacementNode;
       }
 
-      if (successorNode !== removeNode) {
-        removeNode.value = successorNode.value;
-        removeNode = successorNode;
+      if (successorNode !== node) {
+        node.value = successorNode.value;
```

Comment:
I don't recall why I had this one line. The value on the node being removed shouldn't matter. This if condition could be completely removed and we can change this function to only return the successorNode if it's different than the node.

Reply:
Half the tests fail when I try deleting this. Doing some digging to further understand why it's needed. Will include a comment.

Reply:
Think I figured it out.

![Screen Shot 2022-12-31 at 10 49 21 AM](https://user-images.githubusercontent.com/1680910/210153148-7a7f4031-8de5-4aa9-9564-0e2429da8056.png)

`successorNode` is actually the node we are physically removing from the tree, and not `node`. `replacementNode` is the one that will take the old place of `successorNode`. So this line is correct.

I tried recreating the algorithm on my own, and one particular piece I was trying to figure out was how we were guaranteed to delete without creating any orphan nodes. In this implementation, `successorNode` is guaranteed to only have one child, because that's the nature of `findSuccessorNode`. I think that's worth writing in the comments above.

Lastly, looks like `node = successorNode` meant that we weren't returning the successorNode. We were returning the node that was removed. I think we should probably keep it that way and preserve the logic of `removeFixup`, but should definitely be documented.